File: rust/varar-core/tests/conformance_test.rs
//! Port of `ConformanceTest.java` (the var-core half): the var-doc golden gate
//! over every bundle in the shared corpus, plus the registry-projection unit
//! tests. The registry/plan/trace golden gates need per-bundle Rust step
//! fixtures and belong to a future `var` facade crate (as in Java, where they
//! live in the `var` module).

use std::fs;
use std::path::{Path, PathBuf};
use var_core::canonical_json::canonical_stringify;
use var_core::conformance::{parameter_type_names, to_registry_artifact, to_var_doc_artifact};
use var_core::handler::Handler;
use var_core::parse::parse;
use var_core::registry::{add_step, create_registry, define_parameter_type};
use var_core::step_kind::StepKind;
use var_core::value::Value;

fn bundles_dir() -> PathBuf {
    Path::new(env!("CARGO_MANIFEST_DIR")).join("../../conformance/bundles")
}

fn bundle_dirs() -> Vec<PathBuf> {
    let dir = bundles_dir();
    assert!(
        dir.is_dir(),
        "expected conformance corpus at {}",
        dir.display()
    );
    let mut dirs: Vec<PathBuf> = fs::read_dir(&dir)
        .unwrap()
        .filter_map(|e| e.ok().map(|e| e.path()))
        .filter(|p| p.is_dir())
        .collect();
    dirs.sort();
    dirs
}

#[test]
fn var_doc_matches_golden() {
    let mut failures = Vec::new();
    for bundle in bundle_dirs() {
        let name = bundle.file_name().unwrap().to_string_lossy().to_string();
        let source = fs::read_to_string(bundle.join("example.md")).unwrap();
        let doc = parse("example.md", &source);
        let actual = canonical_stringify(&to_var_doc_artifact(&doc));
        let expected = fs::read_to_string(bundle.join("golden").join("var-doc.json")).unwrap();
        if actual != expected {
            failures.push(name);
        }
    }
    assert!(
        failures.is_empty(),
        "var-doc.json mismatch in bundles: {failures:?}"
    );
}

#[test]
fn to_registry_artifact_lists_expressions_and_parsed_parameter_type_names() {
    let r = add_step(
        &create_registry(),
        "I have {int} cukes",
        "s.ts",
        1,
        Handler::noop(),
        None,
    )
    .unwrap();
    let artifact = to_registry_artifact(&r);
    let Value::Map(m) = &artifact else {
        panic!("expected map")
    };
    assert_eq!(Some(&Value::List(vec![])), m.get("parameterTypes"));
    let Value::List(steps) = m.get("steps").unwrap() else {
        panic!("expected steps list")
    };
    assert_eq!(1, steps.len());
    let Value::Map(step0) = &steps[0] else {
        panic!("expected step map")
    };
    assert_eq!(
        Some(&Value::from("I have {int} cukes")),
        step0.get("expression")
    );
    assert_eq!(
        Some(&Value::List(vec![Value::from("int")])),
        step0.get("parameterTypeNames")
    );
}

#[test]
fn to_registry_artifact_reads_parameter_names_from_the_ast_ignoring_escaped_braces() {
    // A naive `{...}` regex would wrongly count the escaped `\{a, b\}`.
    assert_eq!(
        vec!["int".to_string()],
        parameter_type_names("the set \\{a, b\\} has {int} elements")
    );
}

#[test]
fn registry_artifact_projects_custom_parameter_types() {
    let r = create_registry();
    let r = define_parameter_type(
        &r,
        "airport",
        "[A-Z]{3}",
        std::rc::Rc::new(|g: &[&str]| Value::from(g[0])),
    );
    let r = add_step(
        &r,
        "I fly to {airport}",
        "airports.steps",
        1,
        Handler::noop(),
        Some(StepKind::Stimulus),
    )
    .unwrap();
    let artifact = to_registry_artifact(&r);
    let Value::Map(m) = &artifact else {
        panic!("expected map")
    };
    let mut pt = std::collections::BTreeMap::new();
    pt.insert("name".to_string(), Value::from("airport"));
    pt.insert("regexp".to_string(), Value::from("[A-Z]{3}"));
    assert_eq!(
        Some(&Value::List(vec![Value::Map(pt)])),
        m.get("parameterTypes")
    );
    let Value::List(steps) = m.get("steps").unwrap() else {
        panic!("expected steps list")
    };
    let Value::Map(step0) = &steps[0] else {
        panic!("expected step map")
    };
    assert_eq!(
        Some(&Value::List(vec![Value::from("airport")])),
        step0.get("parameterTypeNames")
    );
}

File: java/core/src/test/java/dev/varar/core/ConformanceTest.java
package dev.varar.core;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;
import org.junit.jupiter.api.Named;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;

/**
 * The Milestone 1 conformance gate: for every bundle under the shared, language-neutral
 * {@code conformance/bundles/} corpus, parses {@code example.md}, projects it via {@link
 * Conformance#toVarDocArtifact(Ast.VarDoc)}, serializes with {@link
 * CanonicalJson#canonicalStringify(Object)}, and asserts byte-for-byte equality with the
 * committed {@code golden/var-doc.json}.
 *
 * <p>Port of the var-doc stage of {@code typescript/packages/varar/tests/conformance.test.ts}
 * and {@code python/packages/varar/tests/test_conformance.py::test_var_doc_matches_golden}.
 * Plan/trace stages are later tasks (Milestones 3-4) — this class's golden-driven harness
 * only exercises var-doc; it also carries unit-level (non-golden) coverage of {@link
 * Conformance#toRegistryArtifact}/{@link Conformance#parameterTypeNames}, ported from
 * {@code conformance.test.ts}'s equivalent unit tests — the registry stage's own
 * golden-driven gate lives in {@code dev.varar.ConformanceTest} (the {@code var}
 * module), since it needs a real Java step-definition fixture per bundle, authored against
 * the {@code var} module's {@code Registrar}/{@code StepDefinitions} API.
 *
 * <p>Each bundle is a separately reported {@code @ParameterizedTest} case (not one loop
 * hiding failures behind the first mismatch), keyed by directory name.
 */
class ConformanceTest {

    // Maven runs tests with the module directory (java/core/) as the working
    // directory, so the shared corpus — a sibling of java/, typescript/, python/ at the
    // repo root — is two levels up. Verified empirically: BUNDLES_DIR.toAbsolutePath()
    // resolves to .../conformance/bundles and bundleDirs() finds all 13 bundles.
    private static final Path BUNDLES_DIR = Paths.get("..", "..", "conformance", "bundles");

    // Wrapping each Path in Named<> (rather than returning a bare Stream<Path>) gives every
    // parameterized case its bundle directory name as its JUnit Platform display name — e.g.
    // "08-string-capture" instead of an opaque "[8]" (an IDE/JUnit Console Launcher run
    // renders this; Maven Surefire's own text/XML reports still index by number, but the
    // per-bundle assertion message below names the bundle either way).
    static Stream<Named<Path>> bundleDirs() throws IOException {
        assertTrue(
                Files.isDirectory(BUNDLES_DIR), () -> "Expected conformance corpus at " + BUNDLES_DIR.toAbsolutePath());
        try (Stream<Path> entries = Files.list(BUNDLES_DIR)) {
            return entries
                    .filter(Files::isDirectory)
                    .sorted()
                    .map(dir -> Named.of(dir.getFileName().toString(), dir))
                    .toList()
                    .stream();
        }
    }

    @ParameterizedTest(name = "{0}")
    @MethodSource("bundleDirs")
    void varDocMatchesGolden(Path bundle) throws IOException {
        String source = Files.readString(bundle.resolve("example.md"), StandardCharsets.UTF_8);
        Ast.VarDoc doc = Parse.parse("example.md", source);
        var artifact = Conformance.toVarDocArtifact(doc);
        String actual = CanonicalJson.canonicalStringify(artifact);
        String expected = Files.readString(bundle.resolve("golden").resolve("var-doc.json"), StandardCharsets.UTF_8);
        assertEquals(expected, actual, () -> bundle.getFileName() + "/var-doc.json mismatch");
    }

    private static final Object NOOP_HANDLER = (Runnable) () -> {};

    // Unit-level (not golden-driven) coverage of the registry projection, ported from
    // conformance.test.ts's "toRegistryArtifact lists expressions and parsed
    // parameter-type names" / "... ignoring escaped braces".

    @Test
    void toRegistryArtifactListsExpressionsAndParsedParameterTypeNames() {
        Registry r = Registry.addStep(Registry.createRegistry(), "I have {int} cukes", "s.ts", 1, NOOP_HANDLER, null);

        var artifact = Conformance.toRegistryArtifact(r);
        assertEquals(List.of(), artifact.get("parameterTypes"));
        @SuppressWarnings("unchecked")
        var steps = (List<Object>) artifact.get("steps");
        assertEquals(1, steps.size());
        @SuppressWarnings("unchecked")
        var step = (Map<String, Object>) steps.get(0);
        assertEquals("I have {int} cukes", step.get("expression"));
        assertEquals(List.of("int"), step.get("parameterTypeNames"));
    }

    @Test
    void toRegistryArtifactReadsParameterNamesFromTheAstIgnoringEscapedBraces() {
        // A naive `{...}` regex would wrongly count the escaped `\{a, b\}` as a
        // parameter and yield ["a, b", "int"]; the AST sees only the real {int}.
        Registry r = Registry.addStep(
                Registry.createRegistry(), "the set \\{a, b\\} has {int} elements", "s.ts", 1, NOOP_HANDLER, null);

        assertEquals(
                List.of("int"), Conformance.parameterTypeNames(r.steps().get(0).expression()));
    }

    @Test
    void registryArtifactProjectsCustomParameterTypes() {
        Registry r = Registry.createRegistry();
        r = Registry.defineParameterType(
                r, "airport", java.util.regex.Pattern.compile("[A-Z]{3}"), groups -> groups[0]);
        r = Registry.addStep(r, "I fly to {airport}", "airports.steps", 1, NOOP_HANDLER, StepKind.STIMULUS);
        Map<String, Object> artifact = Conformance.toRegistryArtifact(r);
        assertEquals(List.of(Map.of("name", "airport", "regexp", "[A-Z]{3}")), artifact.get("parameterTypes"));
        assertEquals(
                List.of("airport"), ((Map<?, ?>) ((List<?>) artifact.get("steps")).get(0)).get("parameterTypeNames"));
    }
}
